### ISSUE TITLE
Geometric Aware Loss Function

### ISSUE THREAD

**Issue by alasin** [mbanani/pytorch-clickhere-cnn#1]
Hi @mbanani,

Thanks for releasing the code. I wanted to know how the geometric aware loss function is implemented here. In the paper, authors mention using exponentials of geodesic distance between the viewpoints as a multiplier with the softmax probabilities. But I couldn't find that in your implementation. Could you elaborate on that?

Also, I couldn't understand the reasoning behind bandwidth as it hasn't been mentioned in the paper. Could you elaborate on that too?

Thanks!

**Comment by mbanani** [comment]
Hi @alasin, 

Instead of calculating the exponentials in the loss function, I generate the label using the exponential as shown [here](https://github.com/mbanani/pytorch-clickhere-cnn/blob/83e5411caefbd030d9748bbfebab304e4da7ca2e/utils/util.py#L239). The loss function applied to this label should result in the same loss calculation. I did this to avoid iterating in the loss function.

As for the bandwidth, here's my understanding for why it might be needed; providing a non-zero value for all angles would mean providing a slight reward to all angles regardless how incorrect they are. Through defining a bandwidth, you define the angles that you consider to be partially-correct and deal with all others as false. This is particularly important since the annotations are imperfect estimates and will have some error associated with them. If you have another explanation, I would love to hear it.

I hope my answers are helpful. Let me know if you have any more questions.

**Comment by alasin** [comment]
Thanks for the clarification @mbanani. I get it to some extent but still don't get how -bandwidth to + bandwidth represents the geodesic distance between the two viewpoints.

**Comment by mbanani** [comment]
It does't represent that. It represents some version of acceptable error. Here's an example: 

Consider a similar problem of estimating the rating of an object instead of its pose. If an object's true rating is 3.4 stars, you might be fine with estimating its rating as 3.2 or 3.6, but not higher or lower than that. Hence, you might want to assign a loss of 0 if you estimate 3.4, a low loss value if you estimate 3.5 or 3.3, and the maximum loss for any rating higher than 3.6 or lower than 3.2. 

This is the idea of the bandwidth; it's accepting partially incorrect estimates while penalizing the system with a loss correlated with the estimate's error to some limit, and dealing with all errors beyond that limit as equally incorrect. the -bandwidth to +bandwidth is simply the error you're willing to tolerate. Does this answer your question ?

**Comment by alasin** [comment]
I understand now. It's just that the wording in the paper confused me as they mentioned using geodesic distance. I am implementing the above loss in tensorflow so wanted to clarify everything. Thanks a lot!
